Backend refactoring triggered by 3scale changes

lucasponce opened the pull request:
Work in Progress for:
- Remove 3scale abstraction model in Kiali backend, now 3scale extension will deal directly with Istio Mixer objects: Handlers, Instances and Rules.
- Expose labelSelector to GetIstioConfig endpoint, used mainly to discriminate some Istio config generated by Kiali in 3scale scenario.
- Refactoring cache naming to better alignment with kubernetes package API.
- Expose a new parameter in the config used for 3scale config, this would need a PR in kiali-operator repo.
- Introduce a new handler PATCH /workloads to be used for label workloads (needed to link/unlink 3scale credentials).

It requires https://github.com/kiali/kiali-ui/pull/1833 and https://github.com/kiali/kiali-operator/pull/92

Note that this PR also requires changes on the kiali-operator as it needs access to a new verb (patch) and also it changes the configuration.

Related to https://github.com/kiali/kiali/issues/2860

jotak commented on business/workloads.go:
```diff
@@ -1062,6 +1082,53 @@ func fetchWorkload(layer *Layer, namespace string, workloadName string) (*models
+		kubernetes.JobType,
+		kubernetes.CronJobType,
+		kubernetes.PodType,
+	}
```
If I understand correctly, for a single workload patch request we're going to try to patch every possible kind of workloads with that name?
I guess it works in general because only one type will match with the provided name and the others will return errors / not found?

Is it possible to improve this, and provide the correct type directly from the query params / ui ?
This would give better performances (avoid unnecessary kube queries) and also prevent the corner case where several workloads of different types would have the same name (I guess this corner case would trigger patch errors)

lucasponce replied:
In Kiali we are assuming that we won't have two different controllers with same name in the namespace.
Basically, Istio didn't support that cornercase and we maintained that:
https://github.com/kiali/kiali/blob/master/business/workloads.go#L1065

Yes, perhaps in the handler we could pass an optional parameter with the workloadType and if present to follow that, otherwise fall into fetching all.

But this is a helper, to proper fix this we should revisit the workload model, and I tried to avoid that as it may trigger more major changes that perhaps are better to adress them in a next major version.

xeviknal replied:
I agree with @jotak. It looks like we can easily avoid the multiple calls to k8s API with passing this extra param. I see that the workload model on the backend already have the type field that can be useful in here.
I might be missing something over here though.

lucasponce replied:
Yes, an extra param may help here, but just keeping optional to not break the current model.
In short, kubernetes delegates into the user to not have two different controllers with same name, Istio followed that implicitly in their model and we mapped the workloads using that assumption.
But adding a check for get/update a single workload with a type may help here.
I'll add it during the week to the PR.

What I don't want to open is to introduce any kind of big refactoring as adding constratints into the workload model in Kiali for now, that's something we can discuss in a major iteration.

lucasponce commented:
@xeviknal @jotak I've addressed your comments and added an optional type for workload details (fetch and update flows).
Let me know if it's ok.

I don't remove the "do not merge" label because I'm waiting to finish the UI changes and prepare the PR with the operator.

jotak commented on kubernetes/filters.go:
```diff
@@ -184,3 +184,13 @@ func FilterByRoute(spec map[string]interface{}, protocols []string, service stri
 }
+
+func FilterIstioObjectsForSelector(selector labels.Selector, allObjects []IstioObject) []IstioObject {
+	var istioObjects []IstioObject
```
If the resulting array is to be serialized as json, it's probably safer to initialize this array `istioObjects := []IstioObject{}` ?

lucasponce replied:
It's not exposed but the initialization makes sense

lucasponce commented:
Thanks for the review @xeviknal @hhovsepy @jotak going to merge these ones (backend + frontend) and check with @jmazzitelli if the operator one looks good.
